File: packages/react-native-nitro-filetoolkit/ios/FileMetadataMapper.swift
import Foundation

final class FileMetadataMapper {
  private let fileManager: FileManager

  init(fileManager: FileManager = .default) {
    self.fileManager = fileManager
  }

  func exists(at url: URL) -> Bool {
    (try? fileManager.attributesOfItem(atPath: url.path)) != nil
  }

  func isDirectory(at url: URL) -> Bool {
    let attributes = try? fileManager.attributesOfItem(atPath: url.path)
    return attributes?[.type] as? FileAttributeType == .typeDirectory
  }

  func info(for url: URL) throws -> FileInfo {
    let attributes = try fileManager.attributesOfItem(atPath: url.path)
    let type = attributes[.type] as? FileAttributeType
    let kind: FileKind
    let byteCount: UInt64?
    let symbolicLinkTarget: FileLocation?

    switch type {
    case .typeDirectory:
      kind = .directory
      byteCount = nil
      symbolicLinkTarget = nil
    case .typeSymbolicLink:
      kind = .symbolicLink
      byteCount = nil
      let target = try fileManager.destinationOfSymbolicLink(atPath: url.path)
      let targetURL = target.hasPrefix("/")
        ? URL(fileURLWithPath: target)
        : url.deletingLastPathComponent().appendingPathComponent(target)
      symbolicLinkTarget = FileLocation(
        origin: .uri,
        uri: URL(fileURLWithPath: targetURL.standardizedFileURL.path).absoluteString
      )
    default:
      kind = .file
      byteCount = (attributes[.size] as? NSNumber)?.uint64Value
      symbolicLinkTarget = nil
    }

    return FileInfo(
      kind: kind,
      location: FileLocation(
        origin: .uri,
        uri: URL(fileURLWithPath: url.standardizedFileURL.path).absoluteString
      ),
      name: url.lastPathComponent,
      byteCount: byteCount,
      symbolicLinkTarget: symbolicLinkTarget,
      createdAt: attributes[.creationDate] as? Date,
      modifiedAt: attributes[.modificationDate] as? Date
    )
  }
}

File: packages/react-native-nitro-filetoolkit/android/src/main/java/com/margelo/nitro/filetoolkit/FileMetadataMapper.kt
package com.margelo.nitro.filetoolkit

import android.net.Uri
import android.os.Build
import android.system.ErrnoException
import android.system.Os
import android.system.OsConstants
import java.io.File
import java.time.Instant

internal class FileMetadataMapper {
  fun exists(file: File): Boolean = try {
    Os.lstat(file.path)
    true
  } catch (_: ErrnoException) {
    false
  }

  fun info(file: File): FileInfo {
    val stat = try {
      Os.lstat(file.path)
    } catch (error: ErrnoException) {
      throw FileToolkitException.invalidOperation("location does not exist", error)
    }
    val kind = when {
      OsConstants.S_ISLNK(stat.st_mode) -> FileKind.SYMBOLIC_LINK
      OsConstants.S_ISDIR(stat.st_mode) -> FileKind.DIRECTORY
      else -> FileKind.FILE
    }
    val linkTarget = if (kind == FileKind.SYMBOLIC_LINK) {
      val rawTarget = Os.readlink(file.path)
      val target = if (rawTarget.startsWith('/')) File(rawTarget) else File(file.parentFile, rawTarget)
      uriLocation(target)
    } else {
      null
    }
    val modifiedAt = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      Instant.ofEpochSecond(stat.st_mtime)
    } else {
      null
    }
    return FileInfo(
      kind = kind,
      location = uriLocation(file),
      name = file.name,
      byteCount = if (kind == FileKind.FILE) stat.st_size else null,
      symbolicLinkTarget = linkTarget,
      createdAt = null,
      modifiedAt = modifiedAt,
    )
  }

  private fun uriLocation(file: File): FileLocation {
    val canonical = file.canonicalFile
    return FileLocation(FileLocationOrigin.URI, Uri.fromFile(canonical.absoluteFile).toString())
  }
}
